tests/http/test_fedclient.py

```diff
@@ -211,6 +211,110 @@ def test_client_connect_no_response(self):
         self.assertIsInstance(f.value, RequestSendFailed)
         self.assertIsInstance(f.value.inner_exception, ResponseNeverReceived)
 
+    def test_client_ip_range_blacklist(self):
+        """Ensure that Synapse does not try to connect to blacklisted IPs"""
+        # Set up the ip_range blacklist
+        from netaddr import IPSet
+
+        self.hs.config.federation_ip_range_blacklist = IPSet([
+            "127.0.0.0/8",
+            "fe80::/64",
+        ])
+        self.reactor.lookups["internal"] = "127.0.0.1"
+        self.reactor.lookups["internalv6"] = "fe80:0:0:0:0:8a2e:370:7337"
+        self.reactor.lookups["fine"] = "10.20.30.40"
+        cl = MatrixFederationHttpClient(self.hs, None)
+
+        # Try making a GET request to a blacklisted IPv4 address
+        # ------------------------------------------------------
+        @defer.inlineCallbacks
+        def do_request():
+            with LoggingContext("one") as context:
+                fetch_d = cl.get_json("internal:8008", "foo/bar")
+
+                # Nothing happened yet
+                self.assertNoResult(fetch_d)
+
+                # should have reset logcontext to the sentinel
+                check_logcontext(LoggingContext.sentinel)
+
+                try:
+                    fetch_res = yield fetch_d
+                    defer.returnValue(fetch_res)
+                finally:
+                    check_logcontext(context)
+
+        # Make the request
+        d = do_request()
+        self.pump()
+
+        # Nothing has happened yet
```

Comment:
please remember to test that these deferreds eventually get resolved one way or the other, otherwise we'll end up with requests blocking forever.

Reply:
Still not 100% sure on how to do this. `yield`'ing results in the function timing out (though maybe that's what we want to test?).

Reply:
> `yield`'ing results in the function timing out

yes. The thing to note here is that the test function itself is running *synchronously*: it starts at the beginning, and doesn't return until it is finished. Note that is different to a `Deferred`-returning function, which only does part of its work synchronously, and leaves the rest to be done later (at which point, the `Deferred` will resolve).

Now, the way that we test the asynchronous code from a synchronous function is something like:

1. run a function that returns a deferred
2. satisfy the condition that the function is waiting for (eg: pretend that some time has passed; connect a TCP socket; spoof some results from the database).
  * Once we satisfy the awaited condition, the second half of the function should complete, and hence resolve its deferred.
3. Check that the deferred has now resolved with the expected result.

Often, all that is needed is to "pretend some time has passed". We do this by using a fake Reactor where we can manipulate its idea of the current time. "Pretending time has passed" is then called "advancing the reactor", and is what `self.pump` does.  

[It may be helpful to note that this means that there are actually *two* reactors in play, since the test runner (`trial`) is a Twisted application which runs on a regular reactor: so we end up with the main reactor which is running `trial` and the test reactor which is running the code under test. But in practice, as long as your test functions are synchronous, you can forget about the `trial` reactor.]

So, long story short: how do we test that the deferreds eventually get resolved? well, somewhere in the test we ought to be able to call `self.successResultOf(d)` or `self.failureResultOf(d, cls)` [iirc]. If that assertion fails, it means that the deferred isn't completing, and that is a bad thing!